### PR TITLE
[CI] Open API Specification and Documentation workflow

### LINKED ISSUE #3033: [CI] Open API Specification and Documentation workflow
#### Desired Behavior
Create a CI/CD workflow for Swagger documentation of Rest APIs

_Note: Issue created to track workflow for #2902_

Two workflows needed to track for REST APIs docs
- Needs to be triggered if changes are made in `./handlers` and `./models`
- One workflow to update `./swagger.yaml` with `make swagger-spec`
- Second workflow to update the data for docs.meshery.io i.e.`./docs/_data/swagger.yml` with `make swagger-docs`

#### Reference:
- [Meshery Build & Release Strategy](https://docs.google.com/document/d/11nAxYtz2SUusCYZ0JeNRrOLIxkgmmbUVWz63MBZV2oE/edit#)

### PR DESCRIPTION
**Pull request by aminoxix**
_Signed-off-by: amino19 <anshumaankrprasad76@gmail.com>_

**Description**
Going to add `swagger.yml` for [CI] Open API Specification and Documentation workflow

This PR fixes #3033 

**Notes for Reviewers**
Stage needs help from reviewers!

**[Signed commits](https://github.com/layer5io/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: .github/workflows/swagger.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  push:
+    branches:
+      - 'master'
+    paths:
+      - 'helpers/**'
+  pull_request:
+    branches:
+      - 'master'
+    paths:
+      - 'helpers/**'
```
Comment (ramrodo) [inline .github/workflows/swagger.yml:12]: Remove this `paths` section to see if it can be executed without this.

Comment (amino19) [reply .github/workflows/swagger.yml]: @ramrodo : Sir, can you pl review this. :slightly_smiling_face:

File: .github/workflows/swagger.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
+name: Open API Specification and Documentation
+on:
+  push:
+    branches:
+      - 'master'
+    paths:
+      - 'helpers/**'
```
Comment (ramrodo) [inline .github/workflows/swagger.yml:7]: Also remove this `paths` section

File: .github/workflows/swagger.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      - 'master'
+  workflow_dispatch:
+    inputs:
+        logLevel:
+          description: 'Log level'
+          required: true
+          default: 'warning'
+
+jobs:
+  golangci:
```
Comment (ramrodo) [inline .github/workflows/swagger.yml:19]: You can remove this whole `golangci` job as it is not needed for swagger

**Review by ramrodo — CHANGES_REQUESTED** [review]
The workflow has not been triggered yet. Hoping these suggestions help

File: .github/workflows/swagger.yml (hunk 1 above)
Comment (ramrodo) [reply .github/workflows/swagger.yml]: Sure. I left a couple of comments more

File: .github/workflows/swagger.yml [hunk 4]
```diff
@@ -0,0 +1,47 @@
+name: Open API Specification and Documentation
+on:
+  push:
+    branches:
+      - 'master'
+  pull_request:
+    branches:
+      - 'master'
```
Comment (ramrodo) [inline .github/workflows/swagger.yml:8]: The workflow is not being triggered yet. Let's force more the conditions. Remove all the `push` and `pull_request` segments and set it to:
```yaml
on: [push, pull_request]
```

as you can see in this commit/PR/line:
https://github.com/layer5io/meshery/pull/3038/files#diff-42244398366244a59eb9bb95d4d4f8f8657297b9e8e5bf3a815f3e8cca374ce1R3

**Review by ramrodo — CHANGES_REQUESTED** [review]

File: .github/workflows/swagger.yml [hunk 5]
```diff
@@ -0,0 +1,34 @@
+name: Open API Specification and Documentation
+on: [push, pull_request]
+jobs:
+  swaggerci:
+    name: swagger-build
+    runs-on: ubuntu-latest
+    steps:
+      - name: Setup go-swagger
+      -  uses: minchao/setup-go-swagger@v1
```
Comment (ramrodo) [inline .github/workflows/swagger.yml:9]:
```suggestion
      - uses: minchao/setup-go-swagger@v1
```

Remove white space

File: .github/workflows/swagger.yml [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+      run: |
+          sudo apt install snapd
+          sudo snap install go-swagger
+          sudo apt update
+      uses: actions/cache@v1
+      with:
+        path: |
+          ./handlers
+          ./models
+        run: swagger generate spec -o ./helpers/swagger.yaml --scan-models
```
Comment (ramrodo) [inline .github/workflows/swagger.yml:29]: Move two spaces to the left of this line and the below ones. They should be at the same level of `with`

**Review by ramrodo — CHANGES_REQUESTED** [review]
Hoping these new comments can trigger the workflow :)

File: .github/workflows/swagger.yml [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    - name: swagger-build
+      uses: actions/checkout@v2
+      run: |
+          sudo apt install snapd
+          sudo snap install go-swagger
+          sudo apt update
+    - name: Swagger-Workflow
+      uses: actions/cache@v1
+      with:
+      - name: Open API Specification
```
Comment (ramrodo) [inline .github/workflows/swagger.yml:27]: This `-name` block should be at the same level as the above `-name` block

File: .github/workflows/swagger.yml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    steps:
+    - name: swagger-build
+      uses: actions/checkout@v2
+      run: |
+          sudo apt install snapd
+          sudo snap install go-swagger
+          sudo apt update
+    - name: Swagger-Workflow
+      uses: actions/cache@v1
+      with:
```
Comment (ramrodo) [inline .github/workflows/swagger.yml:26]: Remove these above 3 lines. They are not needed. I think we don't need to save cache yet.

**Review by ramrodo — CHANGES_REQUESTED** [review]
I think we are close :)

File: .github/workflows/swagger.yml (hunk 8 above)
Comment (ramrodo) [reply .github/workflows/swagger.yml:26]: So, besides we don't need them, they are badly indented. For that reason the workflow is failing

**Comment by ramrodo** [comment]
There are still some errros:

image

Please add the `Github Actions` extension for VSCode to see if you have lint errors. You need to fix these errors:
- `path` is not allowed
- Indentation of `with` and `run` are wrong.

**Comment by amino19** [comment]
@ramrodo : Ah, Thanks for this massive tricks! :smile: 
I will commit next, after fixing bug.

File: .github/workflows/swagger.yml [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    - name: swagger-build
+      uses: actions/checkout@v2
+      run: |
+          sudo apt install snapd
+          sudo snap install go-swagger
+          sudo apt update
+    - name: swagger-spec
+      run: swagger generate spec -o ./helpers/swagger.yaml --scan-models
+      uses: actions/cache@v1
+      with:
```
Comment (ramrodo) [inline .github/workflows/swagger.yml:27]: There is still a lint error:

image

Remove:
- `uses: actions/cache@v1` and
- `with:`

they are not needed.

Comment (amino19) [reply .github/workflows/swagger.yml]: @ramrodo : Oh Yes, I forgot we doesn't need to store `cache`!
Thanks for reminder, it's done! :slightly_smiling_face:

File: .github/workflows/swagger.yml [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        with:
+          version: v0.26.1
+
+  build:
+    needs: [swaggerci]
+    name: Swagger-ci
+    runs-on: ubuntu-latest
+    steps:
+    - name: swagger-build
+      uses: actions/checkout@v2
```
Comment (ramrodo) [inline .github/workflows/swagger.yml:19]: This `actions/checkout@v2` is in a bad place. It should be in a single step. See:
https://github.com/meshery/meshery/blob/master/.github/workflows/ci.yml#L27

File: .github/workflows/swagger.yml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    steps:
+    - name: swagger-build
+      run: |
+          sudo apt install snapd
+          sudo snap install go-swagger
+          sudo apt update
+    - name: swagger-spec
+      run: swagger generate spec -o ./helpers/swagger.yaml --scan-models
+    - name: swagger-run:swagger-spec
+      run: swagger serve ./helpers/swagger.yaml
```
Comment (piyushsingariya) [inline .github/workflows/swagger.yml:27]: Remove this. This isn't required.

File: .github/workflows/swagger.yml [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  build:
+    needs: [swaggerci]
+    name: Swagger-ci
+    runs-on: ubuntu-latest
+    steps:
+    - name: swagger-build
+      run: |
+          sudo apt install snapd
+          sudo snap install go-swagger
+          sudo apt update
```
Comment (piyushsingariya) [inline .github/workflows/swagger.yml:23]: Remove this too.

File: .github/workflows/swagger.yml [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  swaggerci:
+    name: swagger-build
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Setup go-swagger
+        uses: minchao/setup-go-swagger@v1
+        with:
+          version: v0.26.1
+
```
Comment (piyushsingariya) [inline .github/workflows/swagger.yml:13]: Paste the other two jobs here.

**Review by piyushsingariya — CHANGES_REQUESTED** [review]

File: .github/workflows/swagger.yml [hunk 14]
```diff
@@ -0,0 +1,16 @@
+name: Open API Specification and Documentation
+on: [push, pull_request]
+jobs:
+  swaggerci:
+    name: swagger-build
```
Comment (piyushsingariya) [inline .github/workflows/swagger.yml:5]:
```suggestion
    name: swagger-docs

```

**Comment by amino19** [comment]
@piyushsingariya : Thanks for suggestion, but Its failing <a href="https://github.com/meshery/meshery/pull/3239/checks?check_run_id=3027826580">again</a>! :confused: 
Can you pl check the <a href="https://github.com/meshery/meshery/pull/3239/files">file</a> again?

File: .github/workflows/swagger.yml (hunk 14 above)
Comment (piyushsingariya) [reply .github/workflows/swagger.yml:5]: Can you replace this?

Comment (amino19) [reply .github/workflows/swagger.yml]: Done! :slightly_smiling_face:

**Review by piyushsingariya — APPROVED** [review]
this is done. Thanks for the hard work! @amino19

**Comment by ramrodo** [comment]
Great! The workflow was finally triggered! Only needs to be fixed. Here is the error:
```
classifier: already annotated as response, can't also be "parameters" -
 // swagger:parameters idPostLoadPreferences
Error: Process completed with exit code 1.
```

**Comment by amino19** [comment]
@ramrodo : Yes, I saw that, Can I know, How that's gonna repaired!

**Comment by ramrodo** [comment]
> @ramrodo : Yes, I saw that, Can I know, How that's gonna repaired! confused

Good question. I don't know. We need to figure out :)

**Comment by piyushsingariya** [comment]
this is a valid error. That's because of me. It has been covered in a different PR. so this can be merged.

**Comment by amino19** [comment]
@ramrodo & @piyushsingariya : Sir, I reworked here :slightly_smiling_face: 
Pl review changes!

**Comment by ramrodo** [comment]
> this is a valid error. That's because of me. It has been covered in a different PR. so this can be merged.

@piyushsingariya Could you share with us that PR that fixes this error, please? Also, it useful to keep track in GitHub the reason of why we are ignoring the workflow error.

**Comment by piyushsingariya** [comment]
@ramrodo it is #3368

**Comment by ramrodo** [comment]
@amino19 could you rebase this branch from master? In order to have the fix in your workflow and have all the tests in green

**Comment by amino19** [comment]
@ramrodo : It's done! :slightly_smiling_face:

**Review by ramrodo — APPROVED** [review]
Great! Thanks to both of you for this change :)

**Pull request merged**